frame_info.hpp

```diff
@@ -0,0 +1,47 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (C) 2021, Raspberry Pi (Trading) Ltd.
+ *
+ * frame_info.hpp - Frame info class for libcamera apps
+ */
+#include <array>
+
+#include <libcamera/controls.h>
+
+struct FrameInfo
+{
+        FrameInfo(libcamera::ControlList &ctrls)
+            : exposure_time(0.0), digital_gain(0.0), colour_gains({{0.0f, 0.0f}}),
+            focus(0.0), aelock(false)
+        {
+                if (ctrls.contains(libcamera::controls::ExposureTime))
+                    exposure_time = ctrls.get<int32_t>(libcamera::controls::ExposureTime);
+
+                if (ctrls.contains(libcamera::controls::AnalogueGain))
+                    analogue_gain = ctrls.get(libcamera::controls::AnalogueGain);
+
+                if (ctrls.contains(libcamera::controls::DigitalGain))
+                    digital_gain = ctrls.get(libcamera::controls::DigitalGain);
+
+                if (ctrls.contains(libcamera::controls::DigitalGain))
+                {
+                    libcamera::Span<const float> gains = ctrls.get(libcamera::controls::ColourGains);
+                    colour_gains[0] = gains[0], colour_gains[1] = gains[1];
+                }
+
+                if (ctrls.contains(libcamera::controls::FocusFoM))
+                    focus = ctrls.get(libcamera::controls::FocusFoM);
+
+                if (ctrls.contains(libcamera::controls::AeLocked))
+                    aelock = ctrls.get(libcamera::controls::AeLocked);
+        }
+
+        unsigned int sequence;
+        float exposure_time;
+        float analogue_gain;
+        float digital_gain;
+        std::array<float, 2> colour_gains;
+        float focus;
+        float fps;
+        bool aelock;
+};
```

Comment:
As per comments for the previous commit, I think it might be neater to put the sequence number in the CompletedRequest, and just to pass that in to the constructor.